server/src/main/java/org/opensearch/search/backpressure/SearchBackpressureManager.java

```diff
@@ -0,0 +1,250 @@
+import org.opensearch.common.SuppressForbidden;
+import org.opensearch.common.util.TokenBucket;
+import org.opensearch.monitor.jvm.JvmStats;
+import org.opensearch.search.backpressure.trackers.ResourceUsageTracker;
+import org.opensearch.tasks.CancellableTask;
+import org.opensearch.tasks.Task;
+import org.opensearch.tasks.TaskResourceTrackingService;
+import org.opensearch.tasks.TaskResourceTrackingService.TaskCompletionListener;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.lang.management.ManagementFactory;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.DoubleSupplier;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+
+/**
+ * SearchBackpressureManager is responsible for monitoring and cancelling in-flight search tasks if they are
+ * breaching resource usage limits when the node is in duress.
+ */
+@SuppressForbidden(reason = "OperatingSystemMXBean#getProcessCpuLoad")
+public class SearchBackpressureManager implements Runnable, TaskCompletionListener {
+    private static final Logger logger = LogManager.getLogger(SearchBackpressureManager.class);
+    private static final OperatingSystemMXBean osMXBean = (OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean();
+
+    private final SearchBackpressureSettings settings;
+    private final TaskResourceTrackingService taskResourceTrackingService;
+    private final List<ResourceUsageTracker> trackers;
+
+    private final Streak cpuBreachesStreak = new Streak();
+    private final Streak heapBreachesStreak = new Streak();
+
+    private final AtomicLong completionCount = new AtomicLong();
+    private final AtomicLong cancellationCount = new AtomicLong();
+    private final AtomicLong limitReachedCount = new AtomicLong();
+
+    private final TokenBucket taskCancellationRateLimiter;
+    private final TokenBucket taskCancellationRatioLimiter;
+
+    private final LongSupplier timeNanosSupplier;
+    private final DoubleSupplier cpuUsageSupplier;
+    private final DoubleSupplier heapUsageSupplier;
+
+    public SearchBackpressureManager(
+        SearchBackpressureSettings settings,
+        TaskResourceTrackingService taskResourceTrackingService,
+        ThreadPool threadPool
+    ) {
+        this(
+            settings,
+            taskResourceTrackingService,
+            threadPool,
+            System::nanoTime,
+            osMXBean::getProcessCpuLoad,
+            () -> JvmStats.jvmStats().getMem().getHeapUsedPercent() / 100.0,
+            Collections.emptyList()
+        );
+    }
+
+    public SearchBackpressureManager(
+        SearchBackpressureSettings settings,
+        TaskResourceTrackingService taskResourceTrackingService,
+        ThreadPool threadPool,
+        LongSupplier timeNanosSupplier,
+        DoubleSupplier cpuUsageSupplier,
+        DoubleSupplier heapUsageSupplier,
```

Comment:
Instead of simply taking time, cpu and memory supplier can we instead have trackers and delegate the responsibility of checking whether resource utilization has breached threshold on their respective tracking resource?

Reply:
We have a requirement for two types of trackers:
1. On a node-level to monitor CPU and memory to check if the node is in duress.
2. On a task-level to monitor the resource usage and make cancellation decisions.

(2) is handled using `ResourceUsageTracker` but (1) didn't fit well in that model. We might have to introduce another `NodeResourceUsageTracker` for this, but it felt like an over-generalization initially.

I'm now thinking of introducing another `NodeMonitor` that takes the sole responsibility of checking if the node is in duress. There we can have multiple implementations of `NodeResourceUsageTrackers` and take this responsibility completely away from `SearchBackpressureService`. How does that sound?